osu.Game/Screens/OnlinePlay/Components/ReadyButton.cs

```diff
@@ -41,38 +39,28 @@ private void load(OsuColour colours)
             SelectedItem.BindValueChanged(item => updateSelectedItem(item.NewValue), true);
         }
 
-        private void updateSelectedItem(PlaylistItem item)
-        {
-            hasBeatmap = findBeatmap(expr => beatmaps.QueryBeatmap(expr));
-        }
-
-        private void beatmapUpdated(ValueChangedEvent<WeakReference<BeatmapSetInfo>> weakSet)
-        {
-            if (weakSet.NewValue.TryGetTarget(out var set))
-            {
-                if (findBeatmap(expr => set.Beatmaps.AsQueryable().FirstOrDefault(expr)))
-                    Schedule(() => hasBeatmap = true);
-            }
-        }
-
-        private void beatmapRemoved(ValueChangedEvent<WeakReference<BeatmapSetInfo>> weakSet)
-        {
-            if (weakSet.NewValue.TryGetTarget(out var set))
-            {
-                if (findBeatmap(expr => set.Beatmaps.AsQueryable().FirstOrDefault(expr)))
-                    Schedule(() => hasBeatmap = false);
-            }
-        }
+        private void updateSelectedItem(PlaylistItem _) => Scheduler.AddOnce(updateBeatmapState);
+        private void beatmapUpdated(ValueChangedEvent<WeakReference<BeatmapSetInfo>> _) => Scheduler.AddOnce(updateBeatmapState);
+        private void beatmapRemoved(ValueChangedEvent<WeakReference<BeatmapSetInfo>> _) => Scheduler.AddOnce(updateBeatmapState);
 
-        private bool findBeatmap(Func<Expression<Func<BeatmapInfo, bool>>, BeatmapInfo> expression)
+        private void updateBeatmapState()
         {
             int? beatmapId = SelectedItem.Value?.Beatmap.Value?.OnlineBeatmapID;
             string checksum = SelectedItem.Value?.Beatmap.Value?.MD5Hash;
 
             if (beatmapId == null || checksum == null)
-                return false;
+                return;
+
+            var databasedBeatmap = beatmaps.QueryBeatmap(b => b.OnlineBeatmapID == beatmapId && b.MD5Hash == checksum);
 
-            return expression(b => b.OnlineBeatmapID == beatmapId && b.MD5Hash == checksum) != null;
+            if (databasedBeatmap == null)
+                hasBeatmap = false;
+            else
+            {
+                // DeletePending isn't updated in the beatmap info query above, need to directly query the beatmap set from database as well.
+                var databasedSet = beatmaps.QueryBeatmapSet(s => s.Equals(databasedBeatmap.BeatmapSet));
+                hasBeatmap = databasedSet?.DeletePending == false;
```

Comment:
@peppy so that's what I was talking about, for some reason, when the beatmap is restored from deletion, the `QueryBeatmap` call above gives an outdated value of `DeletePending` (true), while `QueryBeatmapSet` here gives the correct value of `DeletePending` (false).

that's why I currently have a query performed on the beatmap info table and a query performed on the beatmap set table, not sure how to fix.

Reply:
No test for this PR? Looks like it should be pretty low effort (`ReadyButton` already has good coverage otherwise).

Reply:
Testing the bad way (without a test) looks correct at my end?

![image](https://user-images.githubusercontent.com/191335/103512354-e3040e00-4eab-11eb-9be6-cc1617468207.png)

How are you testing / seeing this?

Reply:
I'm talking about the "restore from deletion" case, deleting it updates the value yes, but restoring from deletion doesn't.

Reply:
Have you investigated why?

Reply:
I haven't yet, but I don't think I'll be able to achieve any results, not familiar with this stuff at all currently.

Reply:
Either way, adding test coverage would make testing this a lot easier.